libs/jsonschema/from_type.go

```diff
@@ -0,0 +1,338 @@
+// Fields can be tagged as "internal" to remove them from the generated schema.
+const internalTag = "internal"
+
+// Annotation for bundle fields that have been deprecated.
+// Fields tagged as "deprecated" are omitted from the generated schema.
+const deprecatedTag = "deprecated"
+
+type constructor struct {
+	// Map of typ.PkgPath() + "." + typ.Name() to the schema for that type.
+	// Example key: github.com/databricks/databricks-sdk-go/service/jobs.JobSettings
+	definitions map[string]Schema
+
+	// Map of typ.PkgPath() + "." + typ.Name() to the corresponding type. Used to
+	// track types that have been seen to avoid infinite recursion.
+	seen map[string]reflect.Type
+
+	// The root type for which the schema is being generated.
+	root reflect.Type
+}
+
+// JSON pointers use "/" as a delimiter to represent nested objects. This means
+// we would instead need to use "~1" to represent "/" if we wish to refer to a
+// key in a JSON object with a "/" in it. Instead of doing that we replace "/" with an
+// additional level of nesting in the output map. Thus the $refs in the generated
+// JSON schema can contain "/" without any issues.
+// see: https://datatracker.ietf.org/doc/html/rfc6901
+//
+// For example:
+// {"a/b/c": "value"} is converted to {"a": {"b": {"c": "value"}}}
+// the $ref for "value" would be "#/$defs/a/b/c" in the generated JSON schema.
+func (c *constructor) Definitions() map[string]any {
+	defs := maps.Clone(c.definitions)
+
+	// Remove the root type from the definitions. We don't need to include it in
+	// the definitions because it will be inlined as the root of the generated JSON schema.
+	delete(defs, typePath(c.root))
+
+	if len(defs) == 0 {
+		return nil
+	}
+
+	res := make(map[string]any)
+	for k, v := range defs {
+		parts := strings.Split(k, "/")
+		cur := res
+		for i, p := range parts {
+			// Set the value for the last part.
+			if i == len(parts)-1 {
+				cur[p] = v
+				break
+			}
+
+			// For all but the last part, create a new map value to add a level
+			// of nesting.
+			if _, ok := cur[p]; !ok {
+				cur[p] = make(map[string]any)
+			}
+			cur = cur[p].(map[string]any)
+		}
+	}
+
+	return res
+}
+
+// FromType converts a reflect.Type to a jsonschema.Schema. Nodes in the final JSON
+// schema are guaranteed to be one level deep, which is done using defining $defs
+// for every Go type and referring them using $ref in the corresponding node in
+// the JSON schema.
+//
+// fns is a list of transformation functions that will be applied in order to all $defs
+// in the schema.
+func FromType(typ reflect.Type, fns []func(typ reflect.Type, s Schema) Schema) (Schema, error) {
+	c := constructor{
+		definitions: make(map[string]Schema),
+		seen:        make(map[string]reflect.Type),
+		root:        typ,
+	}
+
+	_, err := c.walk(typ)
+	if err != nil {
+		return Schema{}, err
+	}
+
+	for _, fn := range fns {
+		for k, v := range c.definitions {
+			c.definitions[k] = fn(c.seen[k], v)
```

Comment:
Is `v` a pointer value -- i.e. can each one of the transformation functions be applied in order without erasing what the previous function applied?

Nevermind: I see the functions are the outer loop. It would be clearer that the value is both an input and output by looping only on the keys of `c.definitions` and using `c.definitions[k]` instead of `v`.